starknet_py/compile/compiler.py

```diff
@@ -0,0 +1,127 @@
+import os
+from pathlib import Path
+from typing import List, NewType, Optional, Tuple, Union
+
+from starkware.starknet.services.api.contract_definition import ContractDefinition
+from starkware.cairo.lang.compiler.constants import MAIN_SCOPE, LIBS_DIR_ENVVAR
+from starkware.cairo.lang.cairo_constants import DEFAULT_PRIME
+from starkware.cairo.lang.compiler.cairo_compile import (
+    get_module_reader,
+)
+from starkware.cairo.lang.compiler.preprocessor.preprocess_codes import preprocess_codes
+from starkware.starknet.compiler.compile import assemble_starknet_contract
+from starkware.starknet.compiler.starknet_pass_manager import starknet_pass_manager
+
+CairoSourceCode = NewType("CairoSourceCode", str)
+CairoFilename = NewType("CairoFilename", str)
+StarknetCompilationSource = NewType(
+    "CairoSource", Union[CairoSourceCode, List[CairoFilename]]
+)
+
+
+class Compiler:
+    """
+    Class for compiling Cairo contracts
+    """
+
+    @staticmethod
+    def create_contract_definition(
+        compiled_contract: Optional[str] = None,
+        contract_source: Optional[StarknetCompilationSource] = None,
+        search_paths: Optional[List[str]] = None,
+    ) -> ContractDefinition:
+        """
+        Creates ContractDefinition either from already compiled contract or contract source code
+
+        :raises ValueError: when neither contract_source and compiled_contract are given
+        :param compiled_contract: an already compiled contract
+        :param contract_source: string containing source code or a list of source files paths
+        :param search_paths: a ``list`` of paths used by starknet_compile to resolve dependencies within contracts
+        :return: a ContractDefinition
+        """
+        if not contract_source and not compiled_contract:
+            raise ValueError(
+                "One of compiled_contract or compilation_source is required."
+            )
+
+        if not compiled_contract:
+            compiled_contract = Compiler.compile_contract(
+                compilation_source=contract_source, search_paths=search_paths
+            )
+
+        return ContractDefinition.loads(compiled_contract)
+
+    @staticmethod
+    def compile_contract(
+        compilation_source: StarknetCompilationSource,
+        search_paths: Optional[List[str]] = None,
+    ) -> str:
+        """
+        Compiles a contract and returns it as string
+
+        :param compilation_source: string containing source code or a list of source files paths
+        :param search_paths: a ``list`` of paths used by starknet_compile to resolve dependencies within contracts
+        :return: string of compiled contract
+        """
+        return Compiler._compile(compilation_source, search_paths=search_paths)
+
+    @staticmethod
+    def _load_cairo_source_code(filename: CairoFilename) -> str:
+        source_file = Path(filename)
+
+        if not source_file.is_file():
+            raise TypeError(f"{filename} does not exist")
```

Comment:
Both `TypeError`s that get raised in this method seem more like `ValueError` to me.
`TypeError` is only meant to be used when something literally has the wrong type, like it expects a list instead of a str.

Reply:
Right, this will need to be changed